docs/v1/p4runtime-id-notes.md

```diff
@@ -0,0 +1,483 @@
+# P4Runtime numeric id notes
+
+There are many kinds of numeric identifiers used in the P4Runtime API
+specification.  While they are all described in the specification and
+accompanying Protobuf `.proto` files, this article attempts to collect
+in one place notes on those identifiers, including:
+
++ Which software entity (or person) chooses their values
++ When the values are chosen
++ What is their scope in "space" and time in which they must be unique
+
+This document was written while referring to the P4Runtime API
+specification as of this commit of the Github repository
+https://github.com/p4lang/p4runtime
+
+```
+commit c9cd4af17c2d562d6e4ec760237f16fea9558fcb
+Author: Chris Sommers <31145757+chrispsommers@users.noreply.github.com>
+Date:   Fri Mar 27 13:43:31 2020 -0700
+```
+
+This is after version 1.1.0 of the specification was released, and it
+is expected that version 1.2.0 will be released soon after this time,
+with very few (if any) changes from the version above.
+
+Here is a complete list of Google Protobuf `.proto` files in this
+version of the p4lang/p4runtime repository:
+
+```bash
+$ find . -name '*.proto'
+./proto/p4/v1/p4runtime.proto
+./proto/p4/v1/p4data.proto
+./proto/p4/config/v1/p4info.proto
+./proto/p4/config/v1/p4types.proto
+```
+
+And here are the number of occurrences of `id`, either as a word by
+itself, or with `_` before it, in each of the files above.  These
+restrictions are made to avoid matching against words like `valid` or
+`bidirectional` in the comments.  As of the referenced version of the
+files, it appears that all occurrences of numeric id values are either
+a word by themselves, or have an underscore character just before
+`id`.
+
+```bash
+$ find . -name '*.proto' | xargs egrep -c '(\bid\b|_id\b)'
+./proto/p4/v1/p4runtime.proto:56
+./proto/p4/v1/p4data.proto:0
+./proto/p4/config/v1/p4info.proto:22
+./proto/p4/config/v1/p4types.proto:0
+```
+
+
+# Numeric ids in the P4Info message
+
+These are all defined in the file `proto/p4/config/v1/p4info.proto`.
+
+
+## Numeric ids that are unique within an entire P4Info message
+
+The messages listed below are all of those that are sub-messages of
+the `P4Info` message, and they contain a `Preamble` sub-message.  Each
+`Preamble` message contains an `id` field, as well as several others
+like `name`, `alias`, and `annotations`.  See the `Preamble` message
+definition for the full list.
+
+Section 6.3 "ID Allocation for P4Info Objects" of the P4Runtime API
+specification says:
+
+    The most significant 8 bits of the ID encodes the object type (as
+    per Table 1).  [ ... ]  The remaining 24 bits must be generated in
+    such a way that the resulting IDs must be globally unique in the
+    scope of the P4Info message.
+
+So all of these id values must be globally unique for a single P4Info
```

Comment:
It might be worth saying that the yare unique w.r.t to all IDs of this type, not really globally unique. You can have a table ID that is the same as a field match ID, or role ID, etc.

Reply:
Makes sense.  Maybe introduce a name to describe this "scope", e.g. "top level P4Info object scope"?  There is probably a better name.

Reply:
I have added another commit attempting to address this comment.  I called the new scope the "P4Info top level scope".